rasterio/_io.pyx

```diff
@@ -820,13 +823,168 @@ cdef class DatasetReaderBase(DatasetBase):
         return sample_gen(self, xy, indexes)
 
 
+cdef class MemoryFileBase(object):
+    """Base for a BytesIO-like class backed by an in-memory file."""
+
+    def __init__(self, initial_bytes=b''):
+        """Map bytes to a file in an in-memory filesystem."""
+        cdef const char *path = NULL
+        cdef VSILFILE *vsi_handle = NULL
+
+        if not isinstance(initial_bytes, (bytearray, bytes)):
+            raise ValueError("Initial bytes must be type bytearray or bytes")
+
+        self.name = os.path.join('/vsimem', str(uuid.uuid4()))
+        self._pos = 0
+
+        self._initial_bytes = initial_bytes
+        cdef unsigned char *buffer = self._initial_bytes
+
+        if self._initial_bytes:
+            path_b = self.name.encode('utf-8')
+            path = path_b
+            vsi_handle = VSIFileFromMemBuffer(path, buffer,
+                                              len(self._initial_bytes), 0)
+            if vsi_handle == NULL:
+                raise OSError('failed to map buffer to file')
+            if VSIFCloseL(vsi_handle) != 0:
+                raise OSError('failed to close mapped file handle')
+
+    def check(self):
+        """True if the in-memory file exists"""
+        path_b = self.name.encode('utf-8')
+        cdef const char *path = path_b
+
+        cdef VSILFILE *fp = VSIFOpenL(path, 'r')
+        if fp == NULL:
+            result = False
+        else:
+            VSIFCloseL(fp)
+            result = True
+
+        return result
+
+    def __len__(self):
+        return self.getbuffer().size
+
+    def close(self):
+        """Close MemoryFile and unlink its in-memory file."""
+        path_b = self.name.encode('utf-8')
+        cdef const char *path = path_b
+        VSIUnlink(path)
+        self._pos = 0
+        self._initial_bytes = None
+
+    def read(self, size=-1):
+        """Read size bytes from MemoryFile."""
+        cdef VSILFILE *fp = NULL
+
+        # Return no bytes immediately if the position is at or past the
+        # end of the file.
+        length = len(self)
+        if self._pos >= length:
+            self._pos = length
+            return b''
+
+        if size == -1:
+            size = length - self._pos
+        else:
+            size = min(size, length - self._pos)
+
+        path_b = self.name.encode('utf-8')
+        cdef const char *path = path_b
+
+        fp = VSIFOpenL(path, 'r')
+        if fp == NULL:
+            raise ValueError("NULL file")
+
+        if VSIFSeekL(fp, self._pos, 0) < 0:
+            raise IOError(
+                "Failed to seek to offset %s in %s.", self._pos, self.name)
+
+        cdef unsigned char *buffer = <unsigned char *>CPLMalloc(size)
+        objects_read = VSIFReadL(buffer, 1, size, fp)
+        VSIFCloseL(fp)
+
+        cdef bytes result = <bytes>buffer[:objects_read]
+        CPLFree(buffer)
+
+        self._pos += len(result)
+        return result
+
+    def seek(self, offset, whence=0):
```

Comment:
`from_what` is used in [file seek](https://docs.python.org/3/tutorial/inputoutput.html); I haven't seen `whence`.  Maybe note in docstring that numbers for it match file seek.